lib/article_json/import/google_doc/html/parser.rb

```diff
@@ -0,0 +1,109 @@
+module ArticleJSON
+  module Import
+    module GoogleDoc
+      module HTML
+        class Parser
+          # @param [String] html
+          def initialize(html)
+            doc = Nokogiri::HTML(html)
+            @body_enumerator = doc.xpath('//body').last.children.to_enum
+
+            css_node = doc.xpath('//head/style').last
+            @css_analyzer = CSSAnalyzer.new(css_node&.inner_text)
+          end
+
+          # Parse the body of the document and return the result
+          # @return [Array]
+          def parsed_content
+            @parsed_content ||= parse_body
+          end
+
+          private
+
+          # Loop over all body nodes and parse them
+          # @return [Array]
+          def parse_body
+            @parsed_content = []
+            while body_has_more_nodes?
+              @parsed_content << begin
+                @current_node = NodeAnalyzer.new(@body_enumerator.next)
+                parse_current_node || next
+              end
+            end
+            @parsed_content
+          end
+
+          # Parse the current node and return an element, if available
+          # @return [Object]
+          def parse_current_node
+            case @current_node.type
+            when :heading then parse_heading
+            when :paragraph then parse_paragraph
+            when :list then parse_list
+            when :image then parse_image
+            when :text_box then parse_text_box
+            when :quote then parse_quote
+            when :hr, :empty, :unknown then nil
```

Comment:
When I have to do method dispatching in ruby, I really miss [clojure multimethods](https://adambard.com/blog/structured-clojure-protocols-and-multimethods/)